adapter-base/src/main/java/org/eclipse/hono/adapter/MessagingClients.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.adapter;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.hono.util.Lifecycle;
+import org.eclipse.hono.util.MessagingType;
+import org.eclipse.hono.util.TenantConstants;
+import org.eclipse.hono.util.TenantObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.healthchecks.HealthCheckHandler;
+
+/**
+ * The set of messaging clients present in a protocol adapter.
+ * <p>
+ * It contains one client for each connected messaging system and allows to get the messaging to be used for a tenant
+ * based on the tenants configuration.
+ */
+public class MessagingClients implements Lifecycle {
+
+    private final Logger log = LoggerFactory.getLogger(MessagingClients.class);
+    private final HashMap<MessagingType, MessagingClient> clients = new HashMap<>();
+
+    /**
+     * Adds a messaging client.
+     *
+     * @param client The client to be added.
+     * @return a reference to this for fluent use.
+     * @throws NullPointerException if the client is {@code null}
+     */
+    public MessagingClients addClient(final MessagingClient client) {
+        Objects.requireNonNull(client);
+        log.info("adding messaging client of type {}", client.getType().name());
+        clients.put(client.getType(), client);
+        return this;
+    }
+
+    /**
+     * Gets the client to be used for messaging.
+     * <p>
+     * The property to determine the messaging type is expected in the {@link TenantConstants#FIELD_EXT_MESSAGING_TYPE}
+     * inside of the {@link TenantConstants#FIELD_EXT} property of the tenant. Valid values are the names of
+     * {@link MessagingType}.
+     *
+     * @param tenant The tenant to select the messaging cleint for.
+     * @return The client that is configured at the tenant or, if this is missing and only one client is set, that one,
+     *         otherwise, the AMQP-based client.
+     * @throws IllegalStateException if no client is set.
+     */
+    public MessagingClient getClientForTenant(final TenantObject tenant) {
+
+        if (clients.isEmpty()) {
+            throw new IllegalStateException("No messaging client present");
```

Comment:
If we consider the fact that no clients have been set a misconfiguration, then we should (also) throw an IllegalStateException in the *start* method.